lib/plugins/game.js

```diff
@@ -1,78 +1,108 @@
+const nbt = require('prismarine-nbt')
 module.exports = inject
 
+const difficultyNames = ['peaceful', 'easy', 'normal', 'hard']
+const gameModes = ['survival', 'creative', 'adventure']
+
 const dimensionNames = {
   '-1': 'minecraft:nether',
   0: 'minecraft:overworld',
   1: 'minecraft:end'
 }
 
-const difficultyNames = ['peaceful', 'easy', 'normal', 'hard']
+function parseGameMode (gameModeBits) {
+  return gameModes[(gameModeBits & 0x03)]
+}
 
-function inject (bot) {
-  bot.game = {}
+function parseHardcore (gameModeBits) {
+  return !!(gameModeBits & 0x04)
+}
 
-  bot._client.on('custom_payload', (packet) => {
-    if (packet.channel === 'minecraft:brand') {
-      bot.game.serverBrand = String.fromCharCode.apply(null, packet.data)
-      bot.emit('game')
-    }
-  })
+function getBrandCustomChannelName (bot) {
+  if (bot.supportFeature('customChannelMCPrefixed')) {
+    return 'MC|Brand'
+  } else if (bot.supportFeature('customChannelIdentifier')) {
+    return 'minecraft:brand'
+  }
+  throw new Error('Unsupported brand channel name')
+}
 
-  bot._client.on('game_state_change', (packet) => {
-    if (packet?.reason === 4 && packet?.gameMode === 1) {
-      bot._client.write('client_command', { action: 0 })
-    }
-  })
+function determineHardcoreFlag (packet) {
+  if (packet.isHardcore !== undefined) {
+    return packet.isHardcore
+  }
+  return parseHardcore(packet.gameMode)
+}
+
+function determineLegacyLevelType (packet) {
+  if (packet.levelType !== undefined) {
+    return packet.levelType
+  }
+  if (packet.isFlat === true) {
+    return 'flat'
+  }
+  return 'default'
+}
+
+function getDimensionNameFromPacket (bot, packet) {
+  if (bot.supportFeature('dimensionIsAnInt')) {
+    return dimensionNames[packet.dimension]
+  } else if (bot.supportFeature('dimensionIsAString')) {
+    return packet.dimension
+  } else if (bot.supportFeature('dimensionIsAWorld')) {
+    return packet.worldName
+  }
+  throw new Error('Unsupported dimension type in login packet')
+}
+
+function handleRespawnPacketData (bot, packet) {
+  bot.game.levelType = determineLegacyLevelType(packet)
```

Comment:
can you inline these if they only get called once? It makes the code easier to read IMO

Reply:
I could, but I literally moved them out into separate functions because not having version dependent logic in the common code actually makes code easier to understand.

Reply:
this should be inlined indeed